### ISSUE TITLE
How to do file upload? (please add to readme.md)

### ISSUE THREAD

**Issue by RichardJECooke** [request/request-promise#140]
I successfully called my hapi.js server route with Postman uploading a file and it looks like this:

![untitled](https://cloud.githubusercontent.com/assets/5468805/17054724/e5baf3ee-5008-11e6-8f77-78f1dc3b4bc1.png)

Then I tried writing a test to upload a file using request-promise but the "file" arrives like this (incorrect and unusable):

![untitled2](https://cloud.githubusercontent.com/assets/5468805/17054751/18040b60-5009-11e6-957b-d738977dde02.png)

**Comment by KpjComp** [comment]
I don't think this is a request-promise issue, but a npm request that request-promise uses.

But from what I can tell you should be using formData, and not form, and you shouldn't need to set the headers manually.  I also think the json=true is not required.

eg.
```javascript
const options = {
  method: 'POST', 
  uri: testData.url + `uploadprofilephoto`,
  formData: {
    image: fs.createReadStream('/home/rje/photo.jpg')
  }
};
const json: IResponse<string> = await request(options);

**Comment by RichardJECooke** [comment]
Thanks! Changing to `formData` was all that was needed. Do you have a link explaining when to use `form`, `formData`, and `body`? Google fails me.

Please add this example to the readme, it's a very common use case.

**Comment by KpjComp** [comment]
I found the information here, like I said it's part of request, and not request-promise.

https://github.com/request/request

**Comment by analog-nico** [comment]
@KpjComp Thanks for helping out on this one and all the other issues you recently joined in on. I really appreciate that!

Let's keep this issue open until this use case went into the docs.

Btw, there is a new [website](http://request.js.org) in the works that will document `request` and `request-promise` in a much more digestible way. [Suggestions](https://github.com/request/request.github.io) are always welcome!

**Comment by CrackerakiUA** [comment]
And if the receiver is also node.js server, where we can find the file? `req.body` and `req.file` is empty.

**Comment by analog-nico** [comment]
@CrackerakiUA You are using Express, right? You'll need a middleware that parses the form data. See [here](http://stackoverflow.com/questions/41410715/best-way-to-upload-file-using-node-js).

**Comment by CrackerakiUA** [comment]
Thank you

**Comment by mrvisser** [comment]
I need guaranteed ordering of parameters in the form post (the multipart file upload should be the final parameter). To achieve that I'm assuming you would have to use the form:

```javascript
const req = request.post('http://my.url/upload');
const form = req.form();
form.append('myparam1', 'myval')
form.append('file', fs.createReadStream('/home/mrvisser/myfile.png'));
```

However when using request-promise, isn't `request.post` going to return a promise instead of the request object?

How is this scenario handled in request promise?

Definite +1 from me some more docs around this.

**Comment by analog-nico** [comment]
Hi @mrvisser , since the above code works for `request` it will also work for `request-promise`. Just call `req.then(...)` afterwards. (`req` actually is a request object and a promise at the same time.)

**Comment by mrvisser** [comment]
Thanks @analog-nico , I didn't realize they would be mixed in like that.

**Comment by mrvisser** [comment]
@analog-nico I refactored some code to do as you request to the following:

```javascript
const req = require('request-promise');
...
  _fileUpload(notify) {
    // POST to the notify upload url. This should result in a redirect with a location header to which we will post
    const post = req.post(notify.upload_url).then((response) => {
      if (response.statusCode < 300 || response.statusCode >= 400) {
        return Promise.reject(new Error(`Expected status code >=300 and <400, but got ${response.statusCode}`));
      } else {
        return Promise.resolve(response.headers['location']);
      }
    });

    // Apply all upload params and then finally the file stream
    const form = post.form();
    _.each(notify.upload_params, (value, key) => form.append(key, value));
    form.append('file', fs.createReadStream(this.bodyPath));

    return post;
  }
```

And I get the error:

`Unhandled rejection TypeError: post.form is not a function`

**Comment by analog-nico** [comment]
@mrvisser Only `req.post(...)` returns the "mixed" object. `req.post(...).then(...)` returns just the promise for further chaining.

Just do the following change regarding your `post` variable:

``` diff
const req = require('request-promise');
...
  _fileUpload(notify) {
    // POST to the notify upload url. This should result in a redirect with a location header to which we will post
-   const post = req.post(notify.upload_url).then((response) => {
+   const post = req.post(notify.upload_url)
+   const postPromise = post.then((response) => {
      if (response.statusCode < 300 || response.statusCode >= 400) {
        return Promise.reject(new Error(`Expected status code >=300 and <400, but got ${response.statusCode}`));
      } else {
        return Promise.resolve(response.headers['location']);
      }
    });

    // Apply all upload params and then finally the file stream
    const form = post.form();
    _.each(notify.upload_params, (value, key) => form.append(key, value));
    form.append('file', fs.createReadStream(this.bodyPath));

-   return post;
+   return postPromise;
  }
```

**Comment by mrvisser** [comment]
Of course, that was pretty silly of me. Thanks for that!

**Comment by analog-nico** [comment]
@mrvisser Well, the return values of `request-promise` are not very intuitive after all. Have a good day!

**Comment by andrefox333** [comment]
Anyone else have tried a combination of multer memoryStorage with request-promise? 
fs.createReadStream works for me when I use onDiskStorage option but having some issues when I switch over to memoryStorage option in multer.

**Comment by andrefox333** [comment]
If anyone comes across this in the future and needing to upload a buffer stream, this solution worked for me. Following code is for request so just make sure to convert to request-promise properly :)

http://stackoverflow.com/a/43020398/7798791

```js
request({
  url: 'http://example.com',
  method: 'POST',
  formData: {
    'regularField': 'someValue',
    'regularFile': someFileStream,
    'customBufferFile': {
      value: fileBufferData,
      options: {
        filename: 'myfile.bin'
      }
    }
  }
}, handleResponse);
```

**Comment by guyklainer** [comment]
Hi @KpjComp,
I'm getting this error when using the formData param:

RequestError: Error: write after end.

My usage is:
options : {...}
  options.formData = {
        file: fs.createReadStream(FILEPATH)
      };

    return await request(options).then...

**Comment by KpjComp** [comment]
`return await request(options).then...`

Not really sure, but why are you mixing Async/Await with promise.then.  the main reason to use async await it to avoid using thenables.

Maybe try ->
```
let ret = await request(options);
//do what you want with ret
return ret;
```

**Comment by guyklainer** [comment]
OK, thanks. I've changed that.
But the same error is still coming.

**Comment by analog-nico** [comment]
Hi @guyklainer , coincidentally there is a new pull request that I still need to merge with [an example](https://github.com/request/request-promise/pull/207/files) of just that. It uses a few more options that  may resolve your issue. If not please post a more detailed code snippet that allows to reproduce the issue.

**Comment by chrsmrtn81** [comment]
Hi all! I've started a project using Electron and need a way to upload files to a server, which I've been trying to use Request for. I've taken an example from the NPM page and modified it to test on my localhost, as below, and it seems to work because I get a success message in the console, but the file is never moved..??

```
const fs = require('fs');
const request = require('request');

var req = request.post(uploadURL, function (err, resp, body) {
    if (err) {
        console.log(err);
    } else {
        console.log(body);
    }
});

var form = req.form();

form.append('upload', fs.createReadStream("C:/nodejs/dave/assets/img/brand_logos/logo.jpg"), {
    filename: "logo.jpg",
    contentType: 'image/jpeg'
});
```

The message I get in my console is as follows...

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="http://localhost/sandbox/img/">here</a>.</p>
<hr>
<address>Apache/2.4.27 (Win64) PHP/7.0.23 Server at localhost Port 80</address>
</body></html>
```

I'm fairly new to JavaScript and feel I might be missing something pretty silly perhaps, any advice would be really welcome though.
